# 4.9 Questions Part 1

# Import data and libraries

In [2]:
# Import Libraries 
import pandas as pd 
import numpy as np
import os 
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [3]:
# Defining path
path = r'/Users/michaelyeboah/Instacart Basket Analysis'

In [4]:
path

'/Users/michaelyeboah/Instacart Basket Analysis'

In [5]:
#Importing customers.csv dataframe
customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'), index_col = False)

In [6]:
#import orders_prods_merged dataset
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_updated_2.pkl'))

### Question 4. Data Wrangling the new Customer data set

In [7]:
#Checking columns listed in customers.csv dataframe for unnecessary or illogical names
customers.columns

Index(['user_id', 'First Name', 'Surnam', 'Gender', 'STATE', 'Age',
       'date_joined', 'n_dependants', 'fam_status', 'income'],
      dtype='object')

In [8]:
#Checking the size of the dataframe customers.csv
customers.shape

(206209, 10)

In [9]:
#Checking the basic information of customers.csv
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   First Name    194950 non-null  object
 2   Surnam        206209 non-null  object
 3   Gender        206209 non-null  object
 4   STATE         206209 non-null  object
 5   Age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


### Observations: First Name has only 194950 non-null entries.  This would mean we have 206209-194950 empty cells.  Confirm this below with value.counts function. 

In [10]:
#Double Checking and confirming First Name column for missing values as it is the only one with fewer non-null objects
customers['First Name'].value_counts(dropna = False)

NaN        11259
Marilyn     2213
Barbara     2154
Todd        2113
Jeremy      2104
           ...  
Merry        197
Eugene       197
Garry        191
Ned          186
David        186
Name: First Name, Length: 208, dtype: int64

### Observations: Confirmed missing 11259 first names, roughly 5.5%, however, it is impossible to determine what their names would be without consulting the source of information. This shouldn't affect any statistics, so we can leave them blank, at least for now.

In [11]:
#Checking the descriptive statistics of customers.csv
customers.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [12]:
#Checking the first ten rows of customers.csv 
customers.head(10)

user_id First Name    Surnam  Gender           STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
5   133128    Cynthia     Noble  Female        Kentucky   43    1/1/2017   
6   152052      Chris    Walton    Male         Montana   20    1/1/2017   
7   168851     Joseph   Hickman    Male  South Carolina   30    1/1/2017   
8    69965     Jeremy      Vang    Male           Texas   47    1/1/2017   
9    82820      Shawn     Chung    Male        Virginia   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374  
5             2    married   49643  
6             0     single   61746  
7             0     single   63712  
8             1    married  162432  
9             2    married   32072

### Observations: May want to rename Surname to 'Last Name' and make 'STATE' not all caps.  Otherwise, I wouldn't remove any columns just yet

In [13]:
#Rename 'Surname' to Last Name in customers.csv
customers.rename(columns = {'Surnam' : 'Last Name'}, inplace = True)

In [14]:
#Rename 'STATE' to State in customers.csv
customers.rename(columns = {'STATE' : 'State'}, inplace = True)

In [15]:
#Confirm changes to column names
customers.head()

user_id First Name Last Name  Gender       State  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

### Question 5. Complete Data Quality and Consistency Checks

In [16]:
#Check for mixed-types in customers.csv
for col in customers.columns.tolist():
    weird = (customers[[col]].applymap(type) != customers[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (customers[weird]) > 0:
        print(col)

First Name


### Observations: First Name is mixed because it has null values in it.  The rest of the columns are fine.  We can change the data type to String to fix this

In [17]:
#Change the datatype of First Name to a String
customers['First Name'] = customers['First Name'].astype('str')

In [18]:
#Double Checking for Mixed Types
for col in customers.columns.tolist():
    weird = (customers[[col]].applymap(type) != customers[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (customers[weird]) > 0:
        print(col)

In [19]:
#Checking for missing values, though this was already shown above as First Name has missing values.  Confirm after type change
customers.isnull().sum()

user_id         0
First Name      0
Last Name       0
Gender          0
State           0
Age             0
date_joined     0
n_dependants    0
fam_status      0
income          0
dtype: int64

In [21]:
#Find any Duplicate entries in customer.csv and assign them to df_dups
df_dups = customers[customers.duplicated()]

In [22]:
df_dups

Empty DataFrame
Columns: [user_id, First Name, Last Name, Gender, State, Age, date_joined, n_dependants, fam_status, income]
Index: []

### Observations: Confirmed no duplicates in the dataframe customers.csv

### Question 6. Combining the data with the rest of the prepared Instacart data

In [ ]:
#Checking ords_prods_merge 
ords_prods_merge.info()

In [30]:
#Checking the ords_prods_group df size
ords_prods_merge.shape

(32404859, 25)

In [31]:
#Re-checking the customers.csv info
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   First Name    206209 non-null  object
 2   Last Name     206209 non-null  object
 3   Gender        206209 non-null  object
 4   State         206209 non-null  object
 5   Age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [ ]:
#Combining the two df using the user_id column
combined = ords_prods_merge.merge(customers, on = 'user_id')

In [ ]:
#Checking first five rows of combined df
combined.head()

In [ ]:
#Checking size of combined df
combined.shape

In [ ]:
#Checking basic info of combined df to ensure information is correct
combined.info()

### Question 7. Ensure the notebook contains logical titles, section headings and descriptive code comments

### Question 8 Export this new dataframe as a pickle file so you can continue to use it in Part 2

In [ ]:
#Exporting combined customer data set with ords_prods_group to pickle file
combined.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'customer_ords_prods_merged.pkl'))

### Question 9. Save the notebook so that you can send it to your tutor for review